flink-cdc-runtime/src/main/java/org/apache/flink/cdc/runtime/operators/schema/SchemaOperator.java

```diff
@@ -78,29 +133,130 @@ public void setup(
+            Schema originalSchema = cachedSchemas.get(originalEvent.tableId());
+            Schema routedTableSchema = cachedSchemas.get(routedTableId);
+            if (originalSchema.equals(routedTableSchema)) {
+                return ChangeEventUtils.recreateDataChangeEvent(originalEvent, routedTableId);
+            }
+            switch (originalEvent.op()) {
+                case INSERT:
+                    return DataChangeEvent.insertEvent(
+                            routedTableId,
+                            regenerateRecordData(
+                                    originalEvent.after(), originalSchema, routedTableSchema),
+                            originalEvent.meta());
+                case UPDATE:
+                    return DataChangeEvent.updateEvent(
+                            routedTableId,
+                            regenerateRecordData(
+                                    originalEvent.before(), originalSchema, routedTableSchema),
+                            regenerateRecordData(
+                                    originalEvent.after(), originalSchema, routedTableSchema),
+                            originalEvent.meta());
+                case DELETE:
+                    return DataChangeEvent.deleteEvent(
+                            routedTableId,
+                            regenerateRecordData(
+                                    originalEvent.before(), originalSchema, routedTableSchema),
+                            originalEvent.meta());
+                case REPLACE:
+                    return DataChangeEvent.replaceEvent(
+                            routedTableId,
+                            regenerateRecordData(
+                                    originalEvent.after(), originalSchema, routedTableSchema),
+                            originalEvent.meta());
+                default:
+                    throw new IllegalArgumentException(
+                            String.format(
+                                    "Unrecognized operation type \"%s\"", originalEvent.op()));
+            }
+        } catch (Exception e) {
+            throw new IllegalStateException("Unable to fill null for empty columns", e);
+        }
+    }
+
+    private RecordData regenerateRecordData(
+            RecordData recordData, Schema originalSchema, Schema routedTableSchema) {
+        // Regenerate record data
+        List<RecordData.FieldGetter> fieldGetters = new ArrayList<>();
+        for (Column column : routedTableSchema.getColumns()) {
+            String columnName = column.getName();
+            int columnIndex = originalSchema.getColumnNames().indexOf(columnName);
+            if (columnIndex == -1) {
+                fieldGetters.add(new NullFieldGetter());
+            } else {
+                RecordData.FieldGetter fieldGetter =
+                        RecordData.createFieldGetter(
+                                originalSchema.getColumn(columnName).get().getType(), columnIndex);
```

Comment:
Can we cache a tunple of originalSchema and its fieldGetterList? As RecordData.createFieldGetter will need to judge DataType and create new class.